Application.mk

```diff
@@ -163,6 +157,9 @@ endif
 
 install:: $(PROGLIST)
 
+show-objs:
```

Comment:
need add object file into libapps for module case

Reply:
maybe i misunderstood how module thing works.
only MAINOBJ turns into a binary and others (OBJS) go into libapps?
how can it work with KERNEL build?

Reply:
for module build, this is two step process:
1.All source files(except main) compile and archieve into libapps like builtin.
2.Main soruce file compile and link with LDLIBS to form the final image.
The difference from kernel mode and other mode is LDLIBS content:
1.Kernel mode: LDLIBS has all userspace library(libc, libproxy, libapps...), so the binary is totally self contain.
2.Other mode: LDLIBS just contain libapps, so the binary will link with the base image to resovle the symbol at in runtime.
But, I see your patch remove libapps from LDLIBS:
https://github.com/apache/incubator-nuttx-apps/pull/83/files

Reply:
thank you for explanation.

Reply:
are there some sort of circular dependency?
libapps -> SYMTABSRC -> install -> linking with LDLIBS, including libapps

Reply:
The right sequence is:
1.Compile all source files(except main) and archieve into libapps.a
  Note: this is done step by step, with this patch we do it in one step
2.Compile main source file and link with libapps.a to get the executable
  Note:
  a.For non kernel build, the executable still contian the unresolved symbols outside libapps.a(.e.g. printf, open...).
  b.For kernel build, since LDLIBS contain all userspace library, the executable is totally self contained.
  c.NuttX support the explicit dynamic loading, but not implicit dynamic loading now, which mean that the executable must be self contained in kernel mode, but it can call dlopen to load the shared library manully.
3.Generate symbols source file from all exectuable unresolved symbols
Note: For kernel build, symbol table should be empty as note 2.b.
4.Compile and archieve symbols source file into libapps.a
Note: this is required for two reason:
a.binfmt need the symbol table to load the exectutable at runtime
b.symbol table ensure the linker don't remove these unresolved symbols
5.libapps.a link with other libarary to get nuttx.bin or userspace.bin
Note: this just happen for flat and protected build, all exectuable need put into the file system and load through binfmt subsytem in kernel build, no big image blob(nuttx.bin or userspace.bin) exist at all.

Reply:
so, flock in $(APPDIR)/Makefile (step 4) should not be necessary because there should not be concurrent ARCHIVE activities (step 1) at that point?

Reply:
Maybe, but it isn't harm to do the lock like other place, and make the code more consistent.